### PR TITLE
#REQUEST 902 - Automatically load the last open/saved song or user-defined template at startup

### PR DESCRIPTION
**Pull request by joxihan**
NEW SONG : Same behavior as current firmware, boot up on blank song.
TEMPLATE SONG: The current default behavior where a new project is initialized for you.
LAST USED SONG: The last project you loaded on the Deluge.
LAST MODIFIED SONG: The last project you saved on the Deluge.

### CODE REVIEW

**Comment by m-m-adams** [comment]
For formatting you can run ./dbt format to get the exact format config used by our CI

If you're using VSCode and install the advanced formatters extension (should be in our suggested extensions) it'll do it on save for you

**Comment by joxihan** [comment]
> For formatting you can run ./dbt format to get the exact format config used by our CI
> 
> If you're using VSCode and install the advanced formatters extension (should be in our suggested extensions) it'll do it on save for you

Thanks Mark, i didn't know that, just run dbt format and checked in

File: src/definitions_cxx.hpp
Comment (joxihan) [reply src/definitions_cxx.hpp:1074]: Agree as default, let me change that.

**Comment by joxihan** [comment]
updated PR following shapphire-arches comments

**Comment by soymonitus** [comment]
When booting the Deluge in the default template song.. when wanting to save the song it will show name "DEFAULT 2". It would be better if the Default song, when loaded, its name or file path is wiped, so it behaves like a blank song, where clicking Save will show you the next available song slot. Say your last song is named SONG34.xml, then the Deluge should suggest you the name "SONG35"

File: src/deluge/deluge.cpp
Comment (soymonitus) [inline src/deluge/deluge.cpp:471]: Here you can add this after the performLoad() action:
```
			if (startupSongMode == StartupSongMode::TEMPLATE) {
				// Wipe the name so the Save action asks you for a new new
				currentSong->name.clear();
			}
```

**Review by soymonitus — CHANGES_REQUESTED** [review]
A change to make the Template feature work as intended

**Comment by soymonitus** [comment]
It doesn't work with booting from songs loaded/saved inside a subfolder. Dir name or path should also be taken into account.

**Comment by joxihan** [comment]
> It doesn't work with booting from songs loaded/saved inside a subfolder. Dir name or path should also be taken into account.

Good catch I'll write the subfolder support tonight hk time thx for the review

**Comment by soymonitus** [comment]
Now it looks good. But i have no permissions to approve so an admin dev should take a look at it 😀 
Thanks a lot for this feature, really useful

File: src/deluge/deluge.cpp
Comment (seangoodvibes) [inline src/deluge/deluge.cpp:474]: The comment says "new new" - did you mean "new name"?

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/gui/l10n/seven_segment.cpp
Comment (soymonitus) [inline src/deluge/gui/l10n/seven_segment.cpp:341]: since screen is only 4 chars, would you say SONG instead? as we are already in the Defaults submenu, there is enough context, to suppose SONG is going to be a setting for Default Song

File: src/deluge/gui/l10n/seven_segment.cpp
Comment (soymonitus) [inline src/deluge/gui/l10n/seven_segment.cpp:342]: Here I would say NEW

**Review by soymonitus — COMMENTED** [review]
Some label changes

File: src/deluge/deluge.cpp
Comment (joxihan) [reply src/deluge/deluge.cpp:474]: updated, thx sean

File: src/deluge/gui/l10n/seven_segment.cpp
Comment (joxihan) [reply src/deluge/gui/l10n/seven_segment.cpp]: updated thx monitus, i ll play around with the 7seg emulator to have more clue on 7seg limitations

**Review by seangoodvibes — APPROVED** [review]

**Comment by seangoodvibes** [comment]
@joxihan could you add documentation to this feature to these documents please?

1. Community Features document: https://github.com/SynthstromAudible/DelugeFirmware/blob/community/docs/community_features.md

I would recommend added section 3.12 to the document under Section 3. General Improvements

2. Change Log Document: https://github.com/SynthstromAudible/DelugeFirmware/blob/community/CHANGELOG.md

Add a brief description of your change under section c1.2.0

Thanks!

**Comment by joxihan** [comment]
> @joxihan could you add documentation for this feature to these documents please?
> 
> 1. Community Features document: https://github.com/SynthstromAudible/DelugeFirmware/blob/community/docs/community_features.md
> 
> I would recommend added section 3.12 to the document under Section 3. General Improvements
> 
> 2. Change Log Document: https://github.com/SynthstromAudible/DelugeFirmware/blob/community/CHANGELOG.md
> 
> Add a brief description of your change under section c1.2.0
> 
> Thanks!

Documentation added :)

File: src/deluge/model/song/song.cpp
Comment (sapphire-arches) [inline src/deluge/model/song/song.cpp:5800]: This is unsafe, `fullPath`'s storage will be deallocated at the end of the function and the returned pointer will therefore be dangling. You should return the string object itself.

File: src/deluge/model/song/song.cpp
Comment (sapphire-arches) [inline src/deluge/model/song/song.cpp:5811]: memory for `dir` is leaked, should only compute the various `strlen`s once.

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [inline src/deluge/deluge.cpp:480]: Should be a `switch` statement over `startupSongMode`.

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [inline src/deluge/deluge.cpp:465]: This code can go under the `case StartupSongMode::TEMPLATE:`

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [inline src/deluge/deluge.cpp:458]:
```suggestion
	case StartupSongMode::TEMPLATE:
	    [[fallthrough]];
	case StartupSongMode::LASTOPENED:
		[[fallthrough]];
```

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [inline src/deluge/deluge.cpp:483]:
```suggestion
	case StartupSongMode::BLANK:
	    [[fallthrough]];
```

File: src/deluge/deluge.cpp
Comment (joxihan) [reply src/deluge/deluge.cpp:465]: hum no, because there's also the logic to load the DEFAULT.XML song if it does already exist

File: src/deluge/deluge.cpp
Comment (joxihan) [reply src/deluge/deluge.cpp:458]: what's that for ?

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [reply src/deluge/deluge.cpp:465]: Right, but it can fall through that case and then execute the rest of the logic.

File: src/deluge/deluge.cpp
Comment (sapphire-arches) [reply src/deluge/deluge.cpp:458]: Tells the compiler that the fall through is intentional, not the result of forgetting a break;

File: src/deluge/deluge.cpp
Comment (joxihan) [reply src/deluge/deluge.cpp:465]: ok didn't know this trick, let me change this when i'm back

File: src/deluge/deluge.cpp
Comment (joxihan) [reply src/deluge/deluge.cpp:483]: interestingly the compiler does not yell at me even tho i dont set this

File: src/deluge/gui/l10n/seven_segment.cpp
Comment (sapphire-arches) [inline src/deluge/gui/l10n/seven_segment.cpp:343]:
```suggestion
        {STRING_FOR_DEFAULT_UI_DEFAULT_STARTUP_SONG_MODE_TEMPLATE, "TMPL"},
```

**Review by sapphire-arches — APPROVED** [review]
LGTM - I think we should add some safety check so if the default song load crashes we fall back to an init song, but that can be future work. Thanks for the contribution!

**Comment by joxihan** [comment]
> LGTM - I think we should add some safety check so if the default song load crashes we fall back to an init song, but that can be future work. Thanks for the contribution!

Thanks, really my pleasure. I did some sanity check it seems to fallback pretty well in case of issue loading the song, (non existing or faulty file).
What is the next step before merging it ?

**Comment by seangoodvibes** [comment]
> > LGTM - I think we should add some safety check so if the default song load crashes we fall back to an init song, but that can be future work. Thanks for the contribution!
> 
> Thanks, really my pleasure. I did some sanity check it seems to fallback pretty well in case of issue loading the song, (non existing or faulty file). What is the next step before merging it ?

Merging now :)! Welcome to the team :D

**Pull request merged**
